core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala

```diff
@@ -0,0 +1,227 @@
+  * extends IntegrationTestHarness. IntegrationTestHarness creates producers and
+  * consumers, and it extends KafkaServerTestHarness. KafkaServerTestHarness starts
+  * brokers, but first it initializes a ZooKeeper server and client, which happens
+  * in ZooKeeperTestHarness.
+  *
+  * To start brokers when the security protocol is SASL_SSL, we need to set a cluster
+  * ACL, which happens optionally in KafkaServerTestHarness. If the security protocol
+  * is SSL or PLAINTEXT, then the ACL isn't set. The remaining ACLs to enable access
+  * to producers and consumers are set here. To set ACLs, we use AclCommand directly.
+  *
+  * Finally, we rely on SaslSetup to bootstrap and setup Kerberos. We don't use
+  * SaslTestHarness here directly because it extends ZooKeeperTestHarness, and we
+  * would end up with ZooKeeperTestHarness.
+  */
+trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
+  override val producerCount = 1
+  override val consumerCount = 2
+  override val serverCount = 3
+  val numRecords = 1
+  val group = "group"
+  val topic = "e2etopic"
+  val part = 0
+  val tp = new TopicPartition(topic, part)
+  val topicAndPartition = new TopicAndPartition(topic, part)
+
+  override protected lazy val trustStoreFile = Some(File.createTempFile("truststore", ".jks"))
+
+  val topicResource = new Resource(Topic, topic)
+  val groupResource = new Resource(Group, group)
+
+  // Arguments to AclCommand to set ACLs. There are three definitions here:
+  // 1- Provides read and write access to topic
+  // 2- Provides only write access to topic
+  // 3- Provides read access to consumer group
+  def topicAclArgs: Array[String] = Array("--authorizer-properties", 
+                                          s"zookeeper.connect=$zkConnect",
+                                          s"--add",
+                                          s"--topic=$topic",
+                                          s"--operation=Read",
+                                          s"--operation=Write",
+                                          s"--allow-principal=$clientPrincipal")
+  def topicWriteAclArgs: Array[String] = Array("--authorizer-properties", 
+                                               s"zookeeper.connect=$zkConnect",
+                                               s"--add",
+                                               s"--topic=$topic",
+                                               s"--operation=Write",
+                                               s"--allow-principal=$clientPrincipal")
+  def groupAclArgs: Array[String] = Array("--authorizer-properties", 
+                                          s"zookeeper.connect=$zkConnect",
+                                          s"--add",
+                                          s"--group=$group",
+                                          s"--operation=Read",
+                                          s"--allow-principal=$clientPrincipal")
```

Comment:
It might be simpler to use the --producer and --consumer option to grant permission to the producer and the consumer, respectively.